### ISSUE TITLE
multipleOf running into floating point issue

### ISSUE THREAD

**Issue by xjamundx** [sagold/json-schema-library#43]
We have the use case of wanting to enforce numbers having no more than 2 decimal places so we're using the `multipleOf` keyword with `.01` to try to achieve this. Unfortunately it fails in unexpected ways.

There is a well known issue with `multipleOf` keyword when used with floating point numbers:
- https://github.com/json-schema-org/json-schema-spec/issues/312
- https://github.com/python-jsonschema/jsonschema/pull/878

Do you think this is something that this parser would like to fix or should we just use strings with regexes to achieve a similar result?

**Comment by sagold** [comment]
Can you describe the error you get? Basically _multipleOf_ with floats does work as expected. For example the following tests are successful: https://github.com/sagold/json-schema-library/blob/main/test/unit/issues/issue43.multipleOf.float.test.ts.

In general, the number type of javascript does have issues for small decimals and division on decimals. So if you need a more robust solution to restrict decimal length, while still maintaining the number type, I recommend a [custom keyword](https://github.com/sagold/json-schema-library/tree/main#add-custom-keyword-validator), for example _maxDigits: 2_.

```ts
import { Draft07, JsonValidator } from "json-schema-library";

const jsonSchema = new Draft07(mySchema, {
    validateKeyword: {
        // register validator for maxDigits
        maxDigits: maxDigitsValidator as JsonValidator
    },
    // register keyword to numbers (adding maxDigits to the list)
    typeKeywords: {
        number: draft07Config.typeKeywords.number.concat("maxDigits")
    }
});
```

with the validator something like

```ts
const maxDigitsValidator: JsonValidator =   (draft, schema, value: number, pointer) => {
   if (isNaN(value) || isNaN(schema.maxDigits)) {
     return undefined;
   }
   
   const [_, digits] = `${value}`.split('.');
   if (digits && digits.length > schema.maxDigits) {
     return {
       type: 'error',
       code: 'max-digits-errors',
       name: 'MaxDigitsError',
       message: `number should not have mor than ${schema.maxDigits} digits`,
       data: { pointer, value, schema }
     }
   }
}
```

As for the error generatiion it is recommended to follow the [custom error introduction](https://github.com/sagold/json-schema-library/tree/main#custom-errors).

**Comment by xjamundx** [comment]
That's great.I'll consider some of the custom parsing option. It's tricky as we're using a non-JS backend and JS front-end and finding JSON Schema has a lot of peculiarities that make it tough to keep them working exactly the same.

An example where it doesn't work is `1.36`:

```
> var draft =  new Draft07({ type: "number", multipleOf: 0.01 })
undefined
> draft.validate(1.36)
[
  {
    type: 'error',
    name: 'MultipleOfError',
    code: 'multiple-of-error',
    message: 'Expected `1.36` in `#` to be multiple of `0.01`',
    data: { multipleOf: 0.01, value: 1.36, pointer: '#' }
  }
]
>
```

**Comment by xjamundx** [comment]
You can also run something like this and get so many random ones:

```ts
for (let i = 1; i < 100; i++) { let num = `2.${i}`; console.log(num, draft.validate(parseFloat(num))); }
```

```

2.73 []
2.74 [
  {
    type: 'error',
    name: 'MultipleOfError',
    code: 'multiple-of-error',
    message: 'Expected `2.74` in `#` to be multiple of `0.01`',
    data: { multipleOf: 0.01, value: 2.74, pointer: '#' }
  }
]
2.75 []
2.76 [
  {
    type: 'error',
    name: 'MultipleOfError',
    code: 'multiple-of-error',
    message: 'Expected `2.76` in `#` to be multiple of `0.01`',
    data: { multipleOf: 0.01, value: 2.76, pointer: '#' }
  }
]
```

**Comment by sagold** [comment]
Yes, you can see this also in you browser console

```
1.36 % 0.01
> 6.938893903907228e-17
```

floating point arithmetic is a known issue and, to my knowledge, unsolvable. For example: https://github.com/ajv-validator/ajv/issues/84#issuecomment-364795655.

To work around most issues json-schema-library shifts decimals by multiplying it with a constant, but in this case javascript results the following

```
1.36 * 10000
> 13600.000000000002
```

Changing this shift to 

```
1.36 * 100000
136000
```

which you can do by 

```
import { settings } from "json-schema-library";
settings.floatingPointPrecision = 100000;
```

but I would not rely on it and use the regex-pattern approach you mentioned. This would align the validation between the programming languages.

I am very open to other suggestions

**Comment by sagold** [comment]
We could add another keyword to _multipleOf_ to cap precision errors. This is also done by ajv and probably other validators. The workaround could look like something in the following example:

```json
{
  "type": "number",
  "multipleOf": 0.01,
  "multipleOfPrecision": 2
}
```

where _multipleOfPrecision_ is just ignored in your other json-schema implementation.

**Comment by sagold** [comment]
You can test this behaviour by customizing your multipleOf-validator

```tsnew Draft(
  {
    type: "number",
    multipleOf: 0.01,
    multipleOfPrecision: 2
  },
  {
    validateKeyword: {
      multipleOf: (draft, schema, value, pointer) => {
        if (isNaN(schema.multipleOf) || typeof value !== "number") {
          return undefined;
        }

        if (schema.multipleOfPrecision) {
          const val = parseFloat(
            (value * 10000).toFixed(schema.multipleOfPrecision)
          );
          const div = parseFloat(
            (schema.multipleOf * 10000).toFixed(schema.multipleOfPrecision)
          );
          if ((val % div) / 10000 !== 0) {
            return draft.errors.multipleOfError({
              multipleOf: schema.multipleOf,
              value,
              pointer,
              schema
            });
          }
          return undefined;
        }
        // https://github.com/cfworker/cfworker/blob/master/packages/json-schema/src/validate.ts#L1061
        // https://github.com/ExodusMovement/schemasafe/blob/master/src/compile.js#L441
        if (((value * 10000) % (schema.multipleOf * 10000)) / 10000 !== 0) {
          return draft.errors.multipleOfError({
            multipleOf: schema.multipleOf,
            value,
            pointer,
            schema
          });
        }
        // also check https://stackoverflow.com/questions/1815367/catch-and-compute-overflow-during-multiplication-of-two-large-integers
        return undefined;
      }
    }
  }
);
```

**Comment by xjamundx** [comment]
This is so great. Thanks for your thorough explanation and showing us all of this magic, not to mention fixing the "unsolvable" :D

**Comment by xjamundx** [comment]
From my colleague Jemma about your fix

> so smooth it’s ridiculous

🏆

**Comment by sagold** [comment]
😀 Yes, the _unsolvable_ label was set too quickly. Thank you, you are welcome!

So does this solve your issue? Then we could merge this.

**Comment by xjamundx** [comment]
It does yes!

**Comment by sagold** [comment]
This issue is fixed with `json-schema-library@9.0.1`.

Thank you for your feedback!
